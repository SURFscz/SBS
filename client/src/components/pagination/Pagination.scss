@import "../../stylesheets/vars";

section.pagination {
    display: flex;
    font-weight: bolder;

    section.pagination-container {
        margin: 40px 40px 20px auto;
        display: flex;
        font-size: 16px;
        align-content: center;
        align-items: center;
    }

    span.link {
        background-color: white;
        color: $medium-blue;
        border: 1px solid $medium-blue;
        border-radius: 6px;
        padding: 6px 12px;
        display: inline-block;
        margin: 0 8px;
        cursor: pointer;

        &:hover {
            background-color: $blue-hover;
            color: $hover-blue;
        }

        &.chevron {
            padding: 4px;
            display: flex;
        }

        &.current {
            background-color: $medium-blue;
            color: white;
            cursor: default;

        }

        &.dots {
            cursor: default;
            border: none;

            &:hover {
                background-color: white;
                color: $medium-blue;
            }
        }
    }

    span svg {
        width: 24px;
        height: auto;
        fill: $medium-blue;
        margin: auto;
    }

}
