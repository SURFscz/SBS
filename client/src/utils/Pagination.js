export const pageCount = 25;

//https://gist.github.com/kottenator/9d936eb3e4e3c3e02598
export const pagination = (page, totalResults) => {
    const delta = 2,
        left = page - delta,
        right = page + delta + 1,
        range = [],
        rangeWithDots = []
    let l;

    for (let i = 1; i <= totalResults; i++) {
        if ((i === 1 || i === totalResults) || (i >= left && i < right)) {
            range.push(i);
        }
    }

    for (const i of range) {
        if (l) {
            if (i - l === 2) {
                rangeWithDots.push(l + 1);
            } else if (i - l !== 1) {
                rangeWithDots.push('...');
            }
        }
        rangeWithDots.push(i);
        l = i;
    }
    return rangeWithDots;
}
