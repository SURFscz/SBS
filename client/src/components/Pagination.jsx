import {pageCount, pagination} from "../utils/Pagination";
import {ReactComponent as ChevronRight} from "../icons/chevron-right.svg";
import {ReactComponent as ChevronLeft} from "../icons/chevron-left.svg";
import "./Pagination.scss";

export default function Pagination({currentPage, onChange, total}) {

    const nbrPages = Math.ceil(total / pageCount);
    const rangeWithDots = pagination(currentPage, nbrPages);

    if (total <= pageCount) {
        return null;
    }

    const ranges = (nbr, index) => {
        const key = `${nbr}_${index}}}`;
        if (typeof nbr === "string" || nbr instanceof String) {
            return <span key={key} className="link dots">{nbr}</span>
        } else if (nbr === currentPage) {
            return <span key={key} className="link current">{nbr}</span>
        } else {
            return <span key={key} className="link" onClick={() => onChange(nbr)}>{nbr}</span>
        }
    }

    return (
        <section className="pagination">
            <section className="pagination-container">
                {(nbrPages > 1 && currentPage !== 1) &&
                <span className="link chevron" onClick={() => onChange(currentPage - 1)}
                      title="Previous page"
                      aria-label="Previous page">
                    <ChevronLeft/>
                </span>}
                {rangeWithDots.map((nbr, index) => ranges(nbr, index))}
                {(nbrPages > 1 && currentPage !== nbrPages) &&
                <span className="link chevron" onClick={() => onChange(currentPage + 1)} title="Next page"
                      aria-label="Next page">
                    <ChevronRight/>
                </span>}
            </section>
        </section>

    )
}
