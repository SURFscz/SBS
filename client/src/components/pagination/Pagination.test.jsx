import React from "react";
import {render} from "@testing-library/react";
import Pagination from "./Pagination";

describe("Pagination", () => {
    it("matches snapshot with default props", () => {
        const {asFragment} = render(
            <Pagination
                page={1}
                totalPages={5}
                onPageChange={() => {}}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot with custom page and totalPages", () => {
        const {asFragment} = render(
            <Pagination
                page={2}
                totalPages={10}
                onPageChange={() => {}}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot with no pages", () => {
        const {asFragment} = render(
            <Pagination
                page={1}
                totalPages={0}
                onPageChange={() => {}}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
