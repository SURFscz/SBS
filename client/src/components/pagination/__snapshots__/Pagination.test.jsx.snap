// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination matches snapshot with custom page and totalPages 1`] = `
<DocumentFragment>
  <section
    class="pagination"
  >
    <section
      class="pagination-container"
    />
  </section>
</DocumentFragment>
`;

exports[`Pagination matches snapshot with default props 1`] = `
<DocumentFragment>
  <section
    class="pagination"
  >
    <section
      class="pagination-container"
    />
  </section>
</DocumentFragment>
`;

exports[`Pagination matches snapshot with no pages 1`] = `
<DocumentFragment>
  <section
    class="pagination"
  >
    <section
      class="pagination-container"
    />
  </section>
</DocumentFragment>
`;
